### PR TITLE
Lazy Modules Documentation Clarifications

### LINKED ISSUE #53366: Lazy Documentation suggestions
## 📚 Documentation

from https://pytorch.org/docs/master/generated/torch.nn.LazyLinear.html?highlight=lazylinear#torch.nn.LazyLinear:

> A torch.nn.Linear module with lazy initialization.

The first sentence should talk about the use, it's not immediate obvious what lazy initialization means unless you are already familiar.  Something like "...with lazy initialization, useful for automatically inferring the `in_features` parameter of an `nn.Linear` module by running a sample input through the network"

> After construction, networks with lazy modules should first be converted to the desired dtype and placed on the desired device. The lazy modules should then be initialized with one or more “dry runs”. These “dry runs” send inputs of the correct size, dtype, and device through the network and to each one of its lazy modules. After this the network can be used as usual.

why one or more dry runs instead of just 1?  It's not clear.

> A final caveat when using lazy modules is that the order of initialization of a network’s parameters may change, since the lazy modules are always initialized after other modules. This can cause the parameters of a network using lazy modules to be initialized differently than the parameters of a network without lazy modules. For example, if the LazyMLP class defined above had a torch.nn.LazyLinear module first and then a regular torch.nn.Linear second, the second module would be initialized on construction and the first module would be initialized during the first dry run.

it's not clear why I would care about the order of initialization.  I think this is suggesting something about random initialization / determinism, should probably link to the existing determinism page.

> Lazy modules can also load regular torch.nn.Parameter s, which replace their torch.nn.UninitializedParameter s:

maybe say something like "load regular nn.Paramters (i.e. you can serialize/deserialize initialized LazyModules and they will remain initialized)

> Note, however, that lazy modules cannot validate that the shape of parameters they load is correct.

It's not clear what this means.  I think it's saying something like, because the initialization in this case is only done once, the module cannot be used in other contexts?

> After construction, networks with lazy modules should first be converted to the desired dtype and placed on the desired device

Probably useful to explain what's going on here.  Ie. something like, "...placed on the desired device.  This is because lazy modules only perform shape inference -- the usual dtype and device placement behavior applies."

### PR DESCRIPTION
**Pull request by emcastillo**
Fixes #53366

@gchanan @albanD 
Thanks for the feedback. Did a first pass trying to address the concerns in the original issue.

### CODE REVIEW

File: torch/nn/modules/lazy.py [hunk 1]
```diff
@@ -110,11 +114,12 @@ class LazyModuleMixin:
 
     A final caveat when using lazy modules is that the order of initialization of a network's
     parameters may change, since the lazy modules are always initialized after other modules.
-    This can cause the parameters of a network using lazy modules to be initialized differently
-    than the parameters of a network without lazy modules.
     For example, if the LazyMLP class defined above had a :class:`torch.nn.LazyLinear` module
     first and then a regular :class:`torch.nn.Linear` second, the second module would be
     initialized on construction and the first module would be initialized during the first dry run.
+    This can cause the parameters of a network using lazy modules to be initialized differently
+    than the parameters of a network without lazy modules due to random number generation. Check
```
Comment (kmaehashi) [inline torch/nn/modules/lazy.py:121]: Let's make it more clear by saying like: `as the order of parameter initializations, which often depends on random number generator which is stateful, is different`

File: torch/nn/modules/lazy.py [hunk 2]
```diff
@@ -129,8 +134,8 @@ class LazyModuleMixin:
                  ('fc2.bias', tensor([0.0019]))])
 
 
-    Lazy modules can also load regular :class:`torch.nn.Parameter` s,
-    which replace their :class:`torch.nn.UninitializedParameter` s:
+    Lazy modules can load regular nn.Parameters (i.e. you can serialize/deserialize
+    initialized LazyModules and they will remain initialized)
```
Comment (kmaehashi) [inline torch/nn/modules/lazy.py:138]:
```suggestion
    Lazy modules can load regular :class:`torch.nn.Parameter` s (i.e. you can serialize/
    deserialize initialized LazyModules and they will remain initialized):
```

File: torch/nn/modules/lazy.py [hunk 3]
```diff
@@ -159,8 +164,8 @@ class LazyModuleMixin:
                             0.2479,  0.1091]])),
                  ('fc2.bias', tensor([0.0019]))])
 
-    Note, however, that lazy modules cannot validate that the shape of parameters they load is correct.
-
+    Note, however, that the loaded parameters will not be replaced when doing a "dry run" if they are initialized
+    and the input shapes are different. This prevents using initialized modules in different contexts.
```
Comment (kmaehashi) [inline torch/nn/modules/lazy.py:168]: I think whether `the input shapes are different` is not related to the lazy behavior.
How about just saying:
```
Note, however, that the loaded parameters will not be replaced when doing a "dry run" if they are already initialized.
```

File: torch/nn/modules/lazy.py [hunk 4]
```diff
@@ -66,8 +66,12 @@ class LazyModuleMixin:
...
 
     After construction, networks with lazy modules should first
-    be converted to the desired dtype and placed on the desired device.
-    The lazy modules should then be initialized with one or more "dry runs".
+    be converted to the desired dtype and placed on the expected device.
+    This is because lazy modules only perform shape inference so the usual dtype
+    and device placement behavior applies.
+    The lazy modules should then be initialized with one "dry run". However, if the
+    module has branches depending on the inputs, more "dry runs" may be necessary to ensure
+    that all the components in the module are properly initialized.
```
[the remark this answers is not in the document]
Comment (emcastillo) [reply torch/nn/modules/lazy.py:74]: I like this better, thanks!

File: torch/nn/modules/linear.py [hunk 5]
```diff
@@ -181,7 +181,8 @@ def extra_repr(self) -> str:
 
 
 class LazyLinear(LazyModuleMixin, Linear):
-    r"""A :class:`torch.nn.Linear` module with lazy initialization.
+    r"""A :class:`torch.nn.Linear` module that infers the `in_features` parameter by
+    running a simple input through the network.
```
Comment (albanD) [inline torch/nn/modules/linear.py:185]: Not sure about this change. The same thing is mentioned in the paragraph just below. Also not sure what the "simple" here refers to?

**Review by albanD — COMMENTED** [review]
Looks good.

 Note that the doc build is failing due to
```
Mar 08 10:12:08 /opt/conda/lib/python3.6/site-packages/torch/nn/modules/lazy.py:docstring of torch.nn.modules.lazy.LazyModuleMixin:83: WARNING: Inline interpreted text or phrase reference start-string without end-string.
```
Which should be fixed.
After that, you can open the `pytorch_python_doc_build` and check the artifact tab to see the rendered doc to make sure it all looks good.

File: torch/nn/modules/linear.py (hunk 5 above)
Comment (emcastillo) [reply torch/nn/modules/linear.py]: gave it a twist :)

**Review by albanD — APPROVED** [review]
LGTM thanks for taking care of this!

**Pull request closed without merging**